kernel/cpu_local/src/lib.rs

```diff
@@ -0,0 +1,333 @@
+//!
+//! CPU-local variables cannot be used until after a given CPU has been initialized,
+//! i.e., its Local APIC (on x86_64) has been discovered and properly configured.
+//! Currently, the [`init()`] routine in this crate should be invoked by
+//! another init routine from the `per_cpu` crate.
+//!
+//! Note that Rust offers the `#[thread_local]` attribute for thread-local storage (TLS),
+//! but there is no equivalent for CPU-local storage.
+//! On x86_64, TLS areas use the `fs` segment register for the TLS base,
+//! and this crates uses the `gs` segment register for the CPU-local base.
+
+#![no_std]
+#![feature(thread_local)]
+extern crate alloc;
+
+use core::{
+    arch::asm,
+    marker::PhantomData,
+};
+use alloc::collections::{BTreeMap, btree_map::Entry};
+use memory::{MappedPages, PteFlags};
+use preemption::{hold_preemption, PreemptionGuard};
+use spin::Mutex;
+
+#[cfg(target_arch = "x86_64")]
+use x86_64::{registers::model_specific::GsBase, VirtAddr};
+
+
+/// The available CPU-local variables, i.e., fields in `per_cpu::PerCpuData` struct.
+//
+// NOTE: These fields and their offsets must be kept in sync with `per_cpu::PerCpuData`.
+//
+pub enum PerCpuField {
+    CpuId,
+    PreemptionCount,
+    TaskSwitchPreemptionGuard,
+    DropAfterTaskSwitch,
+    TestValue,
+    TestString,
+}
+impl PerCpuField {
+    /// Returns the offset of this field in the `per_cpu::PerCpuData` struct.
+    pub const fn offset(&self) -> usize {
+        match self {
+            Self::CpuId => 8,
+            Self::PreemptionCount => 12,
+            Self::TaskSwitchPreemptionGuard => 16,
+            Self::DropAfterTaskSwitch => 24,
+            Self::TestValue => 32,
+            Self::TestString => 40,
+        }
+    }
+}
+
+
+/// This trait must be implemented for each field in `per_cpu::PerCpuData`.
+///
+/// Use the [`define_per_cpu_field!()`] macro to implement this trait.
+///
+/// ## Safety
+/// This is marked `unsafe` because the implementor must guarantee
+/// that the associated `FIELD` constant is correctly specified.
+/// * For example, the implementation of this trait for `CpuId` must specify
+///   the `FIELD` const as [`PerCpuField::CpuId`],
+///   but we cannot verify that here due to cyclic dependency issues.
+pub unsafe trait CpuLocalField: Sized {
+    const FIELD: PerCpuField;
+
+    // In the future, we will add a `DeadlockPrevention` parameter here
+    // to allow each field to dictate what needs to be temporarily disabled
+    // while accessing this field immutably or mutably.
+    // For example, disabling preemption, interrupts, or nothing.
+
+    // const SIZE_CHECK:  () = assert!(Self::FIELD.size() == size_of::<Self>());
+    // const ALIGN_CHECK: () = assert!(Self::FIELD.align() == align_of::<Self>());
+}
+
+/// A macro used to define a type wrapper for inclusion in the `PerCpuData` struct.
+///
+/// This macro implements the [`CpuLocalField`]
+#[macro_export]
+macro_rules! define_per_cpu_field {
```

Comment:
Can we move this macro into `per_cpu`? No other crate should be using this macro.

Reply:
Agreed, will do.

Reply:
Actually I'll remove it entirely.